python/dlr/counter/publisher.py

```diff
@@ -0,0 +1,54 @@
+import concurrent.futures
+import queue
+import threading
+import logging
+
+from .utils import resturlutils
+from . import config
+
+
+# Singleton class
+class MsgPublisher(object):
+    _stop_processing = False
+    _instance = None
+
+    @staticmethod
+    def get_instance():
+        """return single instance of class"""
+        if MsgPublisher._instance is None:
+            MsgPublisher._instance = MsgPublisher()
+        return MsgPublisher._instance
+
+    def __init__(self):
+        try:
+            self.client = resturlutils.RestUrlUtils()
+            self.record_queue = queue.Queue(maxsize=config.CALL_HOME_PUBLISH_MESSAGE_MAX_QUEUE_SIZE)
+            self.event = threading.Event()
+            # start loop
+            executor = concurrent.futures.ThreadPoolExecutor(max_workers=config.CALL_HOME_MAX_WORKERS_THREADS)
+            executor.submit(self._process_queue)
+            logging.info("msg publisher thread pool execution started")
+        except Exception as e:
+            logging.exception("msg publisher thread pool not started", exc_info=True)
+
+    def send(self, data):
+        try:
+            self.record_queue.put(data)
+            self.event.set()
+        except queue.Full as e:
+            logging.exception("ccm msg publisher queue full")
+        except Exception as e:
+            logging.exception("unable to record messages in queue", exc_info=True)
+
+    def _process_queue(self):
+        while self.event.wait() and not MsgPublisher._stop_processing:
```

Comment:
It takes me awhile, but I pinpoint down the thread was hanging because the deadlock here. 

I cannot determine whether it's because `self.event.wait` or `self.record_queue.get`; however this thread is still running after program exited. This need to be fixed. 

Again, it's reproducible with given `python3 -m pytest tests/python/unittest/test_counter_mgr.py` And you can see program won't stop the execution unless key interrupt exception.